libvips/foreign/vips2jpeg.c

```diff
@@ -628,18 +628,25 @@ write_vips( Write *write, int qfac, const char *profile,
 	if( progressive ) 
 		jpeg_simple_progression( &write->cinfo ); 
 
-	/* Turn off chroma subsampling. Follow IM and do it automatically for
-	 * high Q. 
-	 */
-	if( !force_subsample && (
-	        no_subsample ||
-	        qfac >= 90 )) {
-		int i;
-
-		for( i = 0; i < in->Bands; i++ ) { 
-			write->cinfo.comp_info[i].h_samp_factor = 1;
-			write->cinfo.comp_info[i].v_samp_factor = 1;
-		}
+	switch (subsample_mode) {
+        case VIPS_FOREIGN_JPEG_SUBSAMPLE_ON:
+            break;
+        case VIPS_FOREIGN_JPEG_SUBSAMPLE_AUTO:
+            /* Turn off chroma subsampling. Follow IM and do it automatically for
+            * high Q.
+            */
+            if( !(no_subsample ||
```

Comment:
Now that we have subsample_mode, it seems that the no_subsample parameter is redundant and also confusing. 
If a user doesn't want subsampling, he wouldn't pass subsample_mode=VIPS_FOREIGN_JPEG_SUBSAMPLE_AUTO and no_subsample=1, he should just pass subsample_mode=VIPS_FOREIGN_JPEG_SUBSAMPLE_OFF

Reply:
yes, but I leave it here for backwards compatibility.

Reply:
simplified by setting  subsample_mode=VIPS_FOREIGN_JPEG_SUBSAMPLE_OFF when no_subsample=1